Python: Move PartitionSpec to Pydantic

Fokko opened the pull request:
~~Blocked by https://github.com/apache/iceberg/pull/5124~~

~~This also changes the logic a bit. Instead of passing the schema when initializing the PartitionSpec, we pass it to the `.bind(schema: Schema)` function. This will check the source fields, and initialize the transforms.~~

PartitionSpec as Pydantic for easy json (de)serialization and validation

Rebased against the latest master. Removed the bind method that we solely used to check if the column is there. Probably we have to move this down the line.

Fokko commented on python/pyiceberg/transforms.py:
```diff
@@ -91,7 +91,7 @@ def validate(cls, v: Any):
                 return BucketTransform(num_buckets=BUCKET_PARSER.match(v))
             elif v.startswith(TRUNCATE):
-                return TruncateTransform(width=BUCKET_PARSER.match(v))
+                return TruncateTransform(width=TRUNCATE_PARSER.match(v))
```
Oops!

Fokko commented:
@rdblue Feel like some more Python?

rdblue commented on python/pyiceberg/table/partitioning.py:
```diff
@@ -57,24 +81,33 @@ class PartitionSpec:
-    def __eq__(self, other):
+    spec_id: int = Field(alias="spec-id")
+    fields: Tuple[PartitionField, ...] = Field()
+    last_assigned_field_id: int = Field(alias="last-assigned-field-id", default=_PARTITION_DATA_ID_START)
```
We probably don't need this. It is just a high watermark of the field IDs in this spec. We can just define a property that returns it:

```python
    @property
    def last_assigned_field_id(self):
        return max(pf.field_id for pf in self.fields)
```

rdblue commented on python/tests/table/test_metadata.py:
```diff
@@ -205,13 +207,13 @@ def test_updating_metadata():
 def test_serialize_v1():
     table_metadata = TableMetadataV1(**EXAMPLE_TABLE_METADATA_V1).json()
-    expected = """{"location": "s3://bucket/test/location", "table-uuid": "d20125c8-7284-442c-9aea-15fee620737c", "last-updated-ms": 1602638573874, "last-column-id": 3, "schemas": [{"fields": [{"id": 1, "name": "x", "type": "long", "required": true}, {"id": 2, "name": "y", "type": "long", "required": true, "doc": "comment"}, {"id": 3, "name": "z", "type": "long", "required": true}], "schema-id":  ...
+    expected = """{"location": "s3://bucket/test/location", "table-uuid": "d20125c8-7284-442c-9aea-15fee620737c", "last-updated-ms": 1602638573874, "last-column-id": 3, "schemas": [{"fields": [{"id": 1, "name": "x", "type": "long", "required": true}, {"id": 2, "name": "y", "type": "long", "required": true, "doc": "comment"}, {"id": 3, "name": "z", "type": "long", "required": true}], "schema-id":  ...
```
Serialization should not include `last-partition-id`.

Fokko commented:
@rdblue thanks, I've removed it as a field, and added it as a property:

```python
    @property
    def last_assigned_field_id(self) -> int:
        if self.fields:
            return max(pf.field_id for pf in self.fields)
        return _PARTITION_DATA_ID_START
```
